File: python/198. House Robber.py
from ast import List


class Solution:
    def rob(self, nums: List[int]) -> int:
        # rob current, jump2
        # dont rob current, jump1
        n = len(nums)
        dp = [-1 for i in range(n)]
        return self.helper(nums,0,dp)
    def helper(self, nums, i,dp):
        if i>=len(nums):
            return 0
        elif dp[i] != -1:
            return dp[i]
        robCurr = nums[i] + self.helper(nums,i+2,dp)
        notRobCurr = self.helper(nums,i+1,dp)
        dp[i]=max(robCurr,notRobCurr)
        return dp[i]

File: javascript/198. House Robber.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function (nums) {
  var n = nums.length;
  var dp = Array(n).fill(-1);
  return helper(nums, 0, dp);
};

var helper = (nums, index, dp) => {
  if (index >= nums.length) {
    return 0;
  } else if (dp[index] != -1) {
    return dp[index];
  }
  var currRob = nums[index] + helper(nums, index + 2, dp);
  var currNotRob = helper(nums, index + 1, dp);
  dp[index] = Math.max(currRob, currNotRob);
  return dp[index];
};
